### ISSUE TITLE
Citus may behave improperly for certain combinations of `DISTINCT`, `GROUP BY` and aggregates

### ISSUE THREAD

**Issue by onderkalaci** [citusdata/citus#2017]
See the examples below:

```SQL
CREATE TABLE users_table (user_id int, time timestamp, value_1 int, value_2 int, value_3 float, value_4 bigint);
SELECT create_distributed_table('users_table', 'user_id');
COPY users_table FROM 'users_table.data' WITH CSV;

-- run a query with DISTINCT, however, the output contains duplicate values
SELECT 
  DISTINCT user_id, count(value_2)
FROM 
  users_table
GROUP BY 
  user_id, value_2
ORDER BY 
  2 DESC, 1
LIMIT 10;
 user_id | count 
---------+-------
       4 |     9
       5 |     8
       3 |     6
       2 |     5
       3 |     5
       4 |     5
       **5 |     5**
       **5 |     5**
       2 |     4
       1 |     3

```

Obviously, the same query on a local table produces a different result:
```SQL

SELECT DISTINCT user_id,
                count(value_2)
FROM users_table_local
GROUP BY user_id,
         value_2
ORDER BY 2 DESC,
         1
LIMIT 10;
 user_id | count 
---------+-------
       4 |     9
       5 |     8
       3 |     6
       2 |     5
       3 |     5
       4 |     5
       5 |     5
       2 |     4
       1 |     3
       2 |     3
(10 rows)




`users_table.data`:
```
6,2017-11-23 00:09:44.19812,5,2,0,
4,2017-11-23 00:28:45.060248,4,4,3,
5,2017-11-22 22:31:47.62577,3,1,4,
2,2017-11-23 01:14:27.658529,4,4,4,
1,2017-11-23 17:30:34.635085,3,4,4,
4,2017-11-23 02:14:35.047974,4,4,1,
3,2017-11-23 11:18:53.114408,2,2,0,
3,2017-11-22 18:43:51.450263,1,1,4,
5,2017-11-23 16:48:32.08896,5,2,1,
5,2017-11-23 09:05:08.53142,2,2,2,
2,2017-11-23 06:01:08.148777,2,4,2,
2,2017-11-23 00:19:14.138058,3,4,0,
5,2017-11-23 02:50:32.678074,4,2,4,
5,2017-11-23 14:04:49.464065,4,0,0,
4,2017-11-23 15:32:02.360969,5,3,1,
2,2017-11-23 01:04:26.198826,4,3,4,
2,2017-11-23 12:23:16.53276,1,3,3,
4,2017-11-23 12:59:23.28005,4,1,3,
5,2017-11-22 20:43:18.667473,0,3,2,
5,2017-11-23 00:54:44.192608,1,3,2,
3,2017-11-23 00:15:45.610845,1,1,4,
3,2017-11-23 03:23:24.702501,1,2,5,
5,2017-11-23 09:17:14.659809,3,1,1,
2,2017-11-23 11:48:24.943542,0,5,5,
4,2017-11-23 11:35:24.161296,0,4,1,
4,2017-11-23 07:18:43.255734,3,4,1,
5,2017-11-22 21:02:07.575129,2,0,2,
2,2017-11-23 08:22:22.169158,4,2,5,
5,2017-11-23 00:46:13.498577,3,2,2,
5,2017-11-23 07:47:09.542999,1,4,3,
2,2017-11-23 06:50:30.797805,1,1,1,
3,2017-11-23 11:31:17.403189,4,5,3,
6,2017-11-23 14:43:18.024104,3,2,5,
1,2017-11-23 11:11:24.40789,3,4,0,
4,2017-11-23 06:50:08.101207,2,1,5,
5,2017-11-23 16:28:38.455322,2,5,4,
6,2017-11-22 20:15:53.317797,1,1,1,
4,2017-11-23 10:58:31.388305,0,2,3,
4,2017-11-23 01:55:21.824618,3,1,4,
4,2017-11-22 23:48:11.949567,2,0,0,
4,2017-11-23 07:28:42.537255,3,5,3,
3,2017-11-23 11:41:21.157066,3,5,3,
3,2017-11-23 17:10:35.959913,4,3,1,
1,2017-11-22 22:51:43.132261,4,0,3,
4,2017-11-23 08:38:45.877401,4,1,4,
4,2017-11-23 09:08:27.034145,0,0,3,
1,2017-11-23 09:26:42.145043,1,3,3,
5,2017-11-22 22:10:24.315371,1,2,1,
6,2017-11-23 13:51:16.92838,0,4,2,
2,2017-11-23 13:52:54.83829,3,1,4,
2,2017-11-23 11:47:26.900284,2,0,4,
2,2017-11-23 09:59:08.975079,2,2,4,
2,2017-11-23 03:27:50.327051,2,2,0,
6,2017-11-23 00:07:11.068353,1,1,4,
3,2017-11-22 23:24:32.080584,3,2,5,
1,2017-11-23 11:44:57.515981,4,3,4,
4,2017-11-23 14:57:36.278713,3,1,5,
3,2017-11-23 05:01:44.885505,3,5,4,
3,2017-11-22 20:43:31.008625,1,3,2,
2,2017-11-23 08:49:47.029236,4,2,4,
3,2017-11-23 12:56:49.29191,0,5,1,
1,2017-11-23 17:23:03.441394,5,4,3,
5,2017-11-23 06:45:00.678091,3,2,3,
3,2017-11-23 17:18:51.048758,1,5,5,
6,2017-11-23 01:14:55.769581,0,0,5,
4,2017-11-22 23:59:46.493416,3,1,3,
3,2017-11-23 09:57:41.540228,2,2,3,
4,2017-11-23 06:39:06.287818,3,3,2,
2,2017-11-23 06:23:53.572592,4,4,5,
4,2017-11-22 21:33:03.616802,5,4,1,
5,2017-11-23 09:17:47.706703,2,5,3,
3,2017-11-23 04:01:08.04806,5,5,3,
4,2017-11-23 03:34:40.419294,1,0,4,
6,2017-11-23 11:08:04.244582,2,3,2,
5,2017-11-23 11:33:17.739772,4,3,2,
5,2017-11-22 23:10:42.777699,3,4,5,
5,2017-11-23 10:46:56.036776,3,1,3,
2,2017-11-23 06:56:38.46819,0,1,3,
4,2017-11-23 11:45:39.744961,4,5,4,
5,2017-11-23 14:29:02.557934,2,1,2,
5,2017-11-23 15:55:08.493462,0,3,3,
4,2017-11-23 07:09:37.382372,4,1,2,
4,2017-11-23 13:56:39.967304,0,1,5,
4,2017-11-23 05:42:12.89386,2,3,3,
6,2017-11-23 01:13:50.526322,2,4,1,
2,2017-11-23 09:54:28.13665,0,3,4,
3,2017-11-23 03:52:32.008895,4,2,0,
2,2017-11-22 18:19:49.944985,3,5,1,
6,2017-11-23 10:22:11.02918,5,0,5,
5,2017-11-23 10:15:31.764558,2,2,2,
1,2017-11-23 03:32:50.803031,3,2,1,
5,2017-11-23 06:35:05.166535,5,5,1,
5,2017-11-23 02:09:42.27857,3,2,4,
4,2017-11-23 11:22:36.283022,4,1,4,
6,2017-11-22 23:01:24.82289,2,4,1,
3,2017-11-23 12:43:49.261057,4,3,1,
2,2017-11-23 11:41:04.042936,3,4,1,
5,2017-11-23 10:22:39.468816,3,4,3,
5,2017-11-23 06:36:11.898621,5,1,5,
5,2017-11-23 14:19:45.144019,3,3,0,
3,2017-11-23 06:20:05.854857,1,4,2,
```

**Comment by sumedhpathak** [comment]
Do we have a sense of how quick the fix is? This should be part of 7.3, and back ported as well, given the issue.

**Comment by mtuncer** [comment]
more information on the issue
1 - problem is due to a group by column not being present on the target list
2 - worker query included the absent target column in its target list, and the result from worker query is correct as far as the grouping and target list is concerned.
3 - coordinator query planner checks the topmost plan when creating distinct plan, since it sees aggregate plan at top, it assumes that query result is distinct since group by provides that property. 


Problem lies in item 3. Our assumption is not correct since the group by column is not present in target list, therefore it can not provide distinctness property for existing columns.


Solution:
When creating distinct plan, also check if all group by columns exist in target list as non-junk entries, if so it is safe to assume grouping provides distinctness. If some group by fields are not present, then we need to create distinct plan.
